src/raftstore/store/worker/mod.rs

```diff
@@ -1,86 +1,109 @@
 /// Worker contains all workers that do the expensive job in background.
 
 
-use std::sync::{Arc, RwLock};
+use std::sync::Arc;
 use std::thread::{self, JoinHandle, Builder};
-use std::collections::VecDeque;
-use std::time::Duration;
 use std::io;
 use std::fmt::Display;
+use std::sync::atomic::{AtomicUsize, Ordering};
+use std::sync::mpsc::{self, Sender, Receiver};
 use std::time::Instant;
-
-use util::HandyRwLock;
+use std::result;
 
 const TASK_PEEK_INTERVAL_SECS: u64 = 100;
 
+quick_error! {
+    #[derive(Debug)]
+    pub enum Error {
+        Stopped
+        IoError(e: io::Error) {
+            from()
+            display("{}", e)
+        }
+    }
+}
+
+pub type Result<T> = result::Result<T, Error>;
+
 pub trait Runnable<T: Display> {
     fn run(&mut self, t: T);
 }
 
 /// A worker that can schedule time consuming tasks.
 pub struct Worker<T: Display> {
     name: String,
-    queue: Arc<RwLock<VecDeque<T>>>,
+    counter: Arc<AtomicUsize>,
+    sender: Option<Sender<T>>,
+    receiver: Option<Receiver<T>>,
     handle: Option<JoinHandle<()>>,
-    stopped: Arc<RwLock<bool>>,
 }
 
-impl<T: Display + Send + Sync + 'static> Worker<T> {
+impl<T: Display + Send + 'static> Worker<T> {
     pub fn new(name: String) -> Worker<T> {
+        let (tx, rx) = mpsc::channel();
         Worker {
             name: name,
-            queue: Arc::new(RwLock::new(VecDeque::new())),
+            counter: Arc::new(AtomicUsize::new(0)),
+            sender: Some(tx),
+            receiver: Some(rx),
             handle: None,
-            stopped: Arc::new(RwLock::new(false)),
         }
     }
 
-    pub fn start<R: Runnable<T> + Send + 'static>(&mut self, mut runner: R) -> io::Result<()> {
+    pub fn start<R: Runnable<T> + Send + 'static>(&mut self, mut runner: R) -> Result<()> {
         info!("starting working thread: {}", self.name);
-        let stopped = self.stopped.clone();
-        let queue = self.queue.clone();
+        if self.receiver.is_none() {
```

Comment:
No, here we check whether the background thread started. `receiver` will be none if it has started.